src/v/utils/waiter_queue.h

```diff
@@ -41,14 +44,25 @@ class waiter_queue {
     std::vector<std::unique_ptr<wait_item>> _items;
 
 public:
+    ~waiter_queue() {
+        for (auto& i : _items) {
+            i->p.set_exception(ss::abort_requested_exception());
+        }
+        _items.clear();
+    }
+
     ss::future<> await(T value, ss::abort_source& as) {
         auto item = std::make_unique<wait_item>(value);
         auto fut = item->p.get_future();
 
         auto sub_opt = as.subscribe(
-          [item_ptr = item->weak_from_this()]() noexcept {
+          [this, item_ptr = item->weak_from_this()]() noexcept {
               if (item_ptr) {
                   item_ptr->p.set_exception(ss::abort_requested_exception());
+                  std::erase_if(
+                    _items, [item_ptr](const std::unique_ptr<wait_item>& it) {
+                        return it.get() == item_ptr.get();
```

Comment:
the next improvement would be to store wait_items on an intrusive list rather than in a vector. that way we wouldn't need to do this awkward pointer equality test in order to remove the aborted item.